subprojects/file-watching/src/main/java/org/gradle/internal/watch/registry/impl/HierarchicalFileWatcherUpdater.java

```diff
@@ -48,106 +49,120 @@
  *
  * The root project directories are discovered as included builds are encountered at the start of a build, and then they are removed when the build finishes.
  *
- * This is the lifecycle for the watched root project directories:
- * - During a build, there will be various calls to {@link #updateRootProjectDirectories(Collection)},
+ * This is the lifecycle for the watchable hierarchies:
+ * - During a build, there will be various calls to {@link FileWatcherUpdater#registerWatchableHierarchy(File, SnapshotHierarchy)},
  *   each call augmenting the collection. The watchers will be updated accordingly.
- * - When updating the watches, we watch root project directories or old root project directories instead of
+ * - When updating the watches, we watch watchable hierarchies registered for this build or old watched directories from previous builds instead of
  *   directories inside them.
  * - At the end of the build
- *   - stop watching the root project directories with nothing to watch inside
- *   - remember the current watched root project directories as old root directories for the next build
- *   - remove all non-watched root project directories from the old root directories.
+ *   - stop watching the watchable directories with nothing to watch inside
+ *   - remember the currently watched directories as old watched directories for the next build
+ *   - remove everything which isn't watched from the virtual file system.
  */
 public class HierarchicalFileWatcherUpdater implements FileWatcherUpdater {
     private static final Logger LOGGER = LoggerFactory.getLogger(HierarchicalFileWatcherUpdater.class);
 
-    private final Multimap<String, Path> trackedDirectoriesForSnapshot = HashMultimap.create();
-
     private final Set<Path> watchedHierarchies = new HashSet<>();
+    private final FileWatcher fileWatcher;
 
-    private final Set<Path> knownRootProjectDirectoriesFromCurrentBuild = new HashSet<>();
-    private final Set<Path> watchedRootProjectDirectoriesFromPreviousBuild = new HashSet<>();
-
-    private final FileWatcher watcher;
     private final FileSystemLocationToWatchValidator locationToWatchValidator;
+    private final WatchableHierarchies watchableHierarchies;
+
+    private final Set<Path> watchableHierarchiesFromCurrentBuild = new HashSet<>();
+    private final Set<Path> watchedHierarchiesFromPreviousBuild = new HashSet<>();
 
-    public HierarchicalFileWatcherUpdater(FileWatcher watcher, FileSystemLocationToWatchValidator locationToWatchValidator) {
-        this.watcher = watcher;
+    public HierarchicalFileWatcherUpdater(FileWatcher fileWatcher, FileSystemLocationToWatchValidator locationToWatchValidator, Predicate<String> watchFilter) {
+        this.fileWatcher = fileWatcher;
```

Comment:
Why rename this?

Reply:
So it has the same name as in `NonHierarchicalFileWatcherUpdater`.